src/InfluxDB.Collector/Diagnostics/CollectorLog.cs

```diff
@@ -1,15 +1,37 @@
 ﻿using System;
-using System.IO;
+using System.Collections.Concurrent;
+using InfluxDB.Collector.Util;
 
 namespace InfluxDB.Collector.Diagnostics
 {
     public static class CollectorLog
     {
-        public static TextWriter Out { get; set; } = Console.Error;
+        private static readonly ConcurrentDictionary<Action<string, Exception>, object> ErrorHandlers = new ConcurrentDictionary<Action<string, Exception>, object>();
 
-        public static void WriteLine(string format, params object[] args)
+        /// <summary>
+        /// Registers an error handler. Errors reported may not neccessarily always correspond with an exception. That is to say, when the callback is invoked, the exception may be null.
+        /// To unregister a handler, simply dispose of the IDisposable returned from this method
+        /// The order in which handlers are invoked is not defined
+        /// </summary>
+        /// <param name="errorHandler">A func accepting a string and an exception. Note that exception may be null when invoked</param>
+        /// <returns>An IDispoable which when explicitly disposed, will unnregister the handler</returns>
+        public static IDisposable RegisterErrorHandler(Action<string, Exception> errorHandler)
         {
-            Out?.WriteLine(format, args);
+            ErrorHandlers.TryAdd(errorHandler, errorHandler);
+            return new DisposableAction(() => ErrorHandlers.TryRemove(errorHandler, out object junk));
```

Comment:
There is a race condition here, where it is possible for the error handler to be invoked after disposal. Not sure if this is likely to be an issue, but can be rectified by using a lock - in which case I'd switch over to using a plain old list.